cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/GoalOptimizer.java

```diff
@@ -324,19 +303,27 @@ public AnalyzerState state(MetadataClient.ClusterAndGeneration clusterAndGenerat
    * We do this to avoid duplicate optimization cluster model construction and proposal computation.
    *
    * @param operationProgress to report the job progress.
+   * @param allowCapacityEstimation Allow capacity estimation in cluster model if the requested broker capacity is unavailable.
    */
-  public OptimizerResult optimizations(OperationProgress operationProgress) throws InterruptedException {
+  public OptimizerResult optimizations(OperationProgress operationProgress,
+                                       boolean allowCapacityEstimation) throws InterruptedException {
     LoadMonitorTaskRunner.LoadMonitorTaskRunnerState loadMonitorTaskRunnerState = _loadMonitor.taskRunnerState();
     if (loadMonitorTaskRunnerState == LOADING || loadMonitorTaskRunnerState == BOOTSTRAPPING) {
       throw new IllegalStateException("Cannot get proposal because load monitor is in " + loadMonitorTaskRunnerState + " state.");
     } else if (!_loadMonitor.meetCompletenessRequirements(_requirementsWithAvailableValidWindows)) {
       throw new IllegalStateException("Cannot get proposal because model completeness is not met.");
     }
-    if (!validCachedProposal()) {
-      LOG.debug("Cached best proposal is not usable, Cached generation: {}, current generation: {}. Wait for cache update.",
-                _bestProposal == null ? null : _bestProposal.modelGeneration(),
-                _loadMonitor.clusterModelGeneration());
-      synchronized (_cacheLock) {
+    synchronized (_cacheLock) {
+      if (validCachedProposal() && (allowCapacityEstimation || !_bestProposal.isCapacityEstimated())) {
```

Comment:
It seems if the condition is met, we should return the best proposal directly instead of invalidate the cache.

But even after switching the if and else block, you could still have a mismatched situation where request does not allow estimation, and the re-generated proposal still uses estimation. In that case, the switched code would return a proposal not satisfying the user's estimation requirement.

I'm thinking something like

```
public boolean compatibleWithRequest(allowCapacityEstimation, isCapacityEstimated) {
	return allowCapacityEstimation || !isCapacityEstimated;
}

while (!(validCachedProposal() && compatibleWithRequest(allowCapacityEstimation, _bestProposal.isCapacityEstimated()) )) {
	// we cannot use the current bestProposal, and may need to throw exception or re-generate the bestProposal
	if (!compatibleWithRequest(allowCapacityEstimation, _bestProposal.isCapacityEstimated()) {
		throw exception
	}

	invalidate the cache, and regenerate
}

//by the time we reach here, we know it's safe to use the bestProposal since it's both valid and satisfy the estimation requirement
return bestProposal;
```

Reply:
Great catch!
The initial `if` condition is supposed to be the negation of the current condition -- i.e. `(!validCachedProposal() || (!allowCapacityEstimation && _bestProposal.isCapacityEstimated()))`.

I believe, we should avoid the other changes in the while loop (i.e. `while (!(validCachedProposal() && compatibleWithRequest(allowCapacityEstimation, _bestProposal.isCapacityEstimated()) ))`). This is because, in case the capacity retrieval is unable to avoid capacity estimation, we don't want to be stuck in the while loop. When the best proposal is returned by this function, the calling function already checks if the returned proposals involve estimation or not. In other words, if the cached proposals are estimated, and the user request asks for no estimation, we invalidate the cache and get the capacity again. If the returned result is still doing estimation, we throw an exception at a higher level in call hierarchy.